ql/src/java/org/apache/hadoop/hive/ql/exec/repl/util/FileList.java

```diff
@@ -0,0 +1,181 @@
+  }
+
+  /**
+   * Only add operation is safe for concurrent operations.
+   */
+  public void add(String entry) throws SemanticException {
+    if (thresholdHit && !fileListStreamer.isAlive()) {
+      throw new SemanticException("List is not getting saved anymore to file " + backingFile.toString());
+    }
+    try {
+      cache.put(entry);
+    } catch (InterruptedException e) {
+      throw new SemanticException(e);
+    }
+    if (!thresholdHit && cache.size() >= thresholdPoint) {
+      initStoreToFile(cache.size());
+    }
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (!thresholdHit) {
+      return (cache != null && !cache.isEmpty());
+    }
+    if (nextElement != null) {
+      return true;
+    }
+    if (noMoreElement) {
+      return false;
+    }
+    nextElement = readNextLine();
+    if (nextElement == null) {
+      noMoreElement = true;
+    }
+    return !noMoreElement;
+  }
+
+  @Override
+  public String next() {
+    if (!hasNext()) {
+      throw new NoSuchElementException("No more element in the list backed by " + backingFile);
+    }
+    String retVal = nextElement;
+    nextElement = null;
+    return thresholdHit ? retVal : cache.poll();
+  }
+
+  private synchronized void initStoreToFile(int cacheSize) {
+    if (!thresholdHit) {
+      fileListStreamer.setName(getNextID());
+      fileListStreamer.setDaemon(true);
+      fileListStreamer.start();
+      thresholdHit = true;
+      LOG.info("Started streaming the list elements to file: {}, cache size {}", backingFile, cacheSize);
+    }
+  }
+
+  private String readNextLine() {
+    String nextElement = null;
+    try {
+      if (backingFileReader == null) {
+        FileSystem fs = FileSystem.get(backingFile.toUri(), conf);
+        if (fs.exists(backingFile)) {
+          backingFileReader = new BufferedReader(new InputStreamReader(fs.open(backingFile)));
+        }
+      }
+      nextElement = (backingFileReader == null) ? null : backingFileReader.readLine();
+    } catch (IOException e) {
+      LOG.error("Unable to read list from backing file " + backingFile, e);
+    }
+    return nextElement;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (thresholdHit && fileListStreamer != null) {
+      fileListStreamer.close();
+    }
+    if (backingFileReader != null) {
+      backingFileReader.close();
+    }
+    LOG.info("Completed close for File List backed by:{}, thresholdHit:{} ", backingFile, thresholdHit);
+  }
+
+  private static String getNextID() {
+    if (Integer.MAX_VALUE == fileListStreamerID) {
+      //reset the counter
+      fileListStreamerID = 0;
+    }
+    fileListStreamerID++;
+    return FILE_LIST_STREAMER_PREFIX  + fileListStreamerID;
+  }
+
+  public int getThreshold(int cacheSize) {
+    boolean copyAtLoad = conf.getBoolVar(HiveConf.ConfVars.REPL_DATA_COPY_LAZY);
+    return copyAtLoad ? 0 : (int)(cacheSize * thresholdFactor);
+  }
+
+  @VisibleForTesting
+  public boolean isStreamingToFile() {
```

Comment:
for unit tests you can make them package public

Reply:
Did you mean package-private?

Reply:
Do you need public methods? For tests a package level visibility should be fine

Reply:
It was protected by @VisibleForTesting. But I will remove public also.

Reply:
If you mark it as public it becomes accesible. So its not protected by @VisibleForTesting anymore